src/xmipp/libraries/reconstruction/resolution_fso.cpp

```diff
@@ -25,1334 +25,1078 @@
-			int maskrad = 125;
-			half1 = mono.createDataTest(xdim, ydim, zdim, wavelength, mean, 0.0);
-			half2 = half1;
-
-			mono.addNoise(phalf1, 0, std);
-			mono.addNoise(phalf2, 0, std);
-			FileName fn;
-			Image<double> saveImg;
-			fn = formatString("inputVol1_large.vol");
-			saveImg() = half1;
-			saveImg.write(fn);
-			fn = formatString("inputVol2_large.vol");
-			saveImg() = half2;
-			saveImg.write(fn);
-		}
-		else
+	// Applying the mask
+	if (fnmask!="")
+	{
+		mask.read(fnmask);
+		FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(pmask)
 		{
-			std::cout << "Reading data..." << std::endl;
-			Image<double> imgHalf1, imgHalf2;
-			imgHalf1.read(fnhalf1);
-			imgHalf2.read(fnhalf2);
-
-			half1 = imgHalf1();
-			half2 = imgHalf2();
-
-			if (! fnmask.isEmpty())
-			{
-				mask.read(fnmask);
-				FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(pmask)
-				{
-					double valmask = (double) DIRECT_MULTIDIM_ELEM(pmask, n);
-					DIRECT_MULTIDIM_ELEM(phalf1, n) = DIRECT_MULTIDIM_ELEM(phalf1, n) * valmask;
-					DIRECT_MULTIDIM_ELEM(phalf2, n) = DIRECT_MULTIDIM_ELEM(phalf2, n) * valmask;
-				}
-			}
-			mask.clear();
-			pmask.clear();
+			double valmask = (double) DIRECT_MULTIDIM_ELEM(pmask, n);
+			DIRECT_MULTIDIM_ELEM(half1, n) = DIRECT_MULTIDIM_ELEM(half1, n) * valmask;
+			DIRECT_MULTIDIM_ELEM(half2, n) = DIRECT_MULTIDIM_ELEM(half2, n) * valmask;
 		}
+	}
+	mask.clear();
+	pmask.clear();
 
-		phalf1.setXmippOrigin();
-		phalf2.setXmippOrigin();
 
-    	std::cout << "Starting..." << std::endl;
-    }
+	half1.setXmippOrigin();
+	half2.setXmippOrigin();
 
+	// fftshift must by applied before computing the fft. This will be computed later
+	CenterFFT(half1, true);
+	CenterFFT(half2, true);
+}
 
-    void saveFSCToMetadata(MetaData &mdRes,
-    		const MultidimArray<double> &freq,
-			const MultidimArray<double> &FSC, FileName &fnmd)
-    {
-    	size_t id;
-    	FOR_ALL_ELEMENTS_IN_ARRAY1D(freq)
-		{
-			if (i>0)
-			{
-				id=mdRes.addObject();
-				mdRes.setValue(MDL_RESOLUTION_FREQ,dAi(freq, i),id);
-				mdRes.setValue(MDL_RESOLUTION_FRC,dAi(FSC, i),id);
-				mdRes.setValue(MDL_RESOLUTION_FREQREAL,1./dAi(freq, i),id);
-			}
-		}
-		mdRes.write(fnmd);
-    }
 
-    void saveAnisotropyToMetadata(MetaData &mdAnisotropy,
-    		const MultidimArray<double> &freq,
-			const MultidimArray<double> &anisotropy, FileName &fnmd)
-    {
-    	size_t objId;
-		FOR_ALL_ELEMENTS_IN_ARRAY1D(anisotropy)
+void ProgFSO::saveAnisotropyToMetadata(MetaData &mdAnisotropy,
+		const MultidimArray<double> &freq,
+		const MultidimArray<double> &anisotropy)
+{
+	size_t objId;
+	FOR_ALL_ELEMENTS_IN_ARRAY1D(anisotropy)
+	{
+		if (i>0)
 		{
-			if (i>0)
-			{
-			objId = mdAnisotropy.addObject();
-			mdAnisotropy.setValue(MDL_RESOLUTION_FREQ, dAi(freq, i),objId);
-			mdAnisotropy.setValue(MDL_RESOLUTION_FRC, dAi(anisotropy, i),objId);
-			mdAnisotropy.setValue(MDL_RESOLUTION_FREQREAL, 1.0/dAi(freq, i),objId);
-			}
+		objId = mdAnisotropy.addObject();
```

Comment:
again - add row locally, set the values, add entire row.
Why do you skip the first row anyway?